logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java

```diff
@@ -111,50 +88,48 @@ public long readLastBlock(final DirectBuffer readBuffer) {
    */
   @Override
   public long lookUpApproximateAddress(final long position) {
-    final var low = reader.getFirstIndex();
-    final var high = reader.getLastIndex();
-
     if (position == Long.MIN_VALUE) {
-      final var optionalEntry = findEntry(reader.getFirstIndex());
-      return optionalEntry.map(Indexed::index).orElse(LogStorage.OP_RESULT_INVALID_ADDR);
+      return readFirstBlock();
     }
 
-    // when the log is empty, the last index is defined as first index - 1
-    if (low >= high) {
-      // need a better way to figure out how to know if its empty
-      if (findEntry(low).isEmpty()) {
-        return LogStorage.OP_RESULT_INVALID_ADDR;
-      }
-
-      return low;
+    final long address;
+    try {
+      address = reader.seekToAsqn(position);
+    } catch (final UnsupportedOperationException e) {
+      // in case we seek to an unknown position
+      return LogStorage.OP_RESULT_INVALID_ADDR;
     }
 
-    final var index = zeebeIndexMapping.lookupPosition(position);
-    final long result;
-    if (index == -1) {
-      result = low;
-    } else {
-      result = index;
+    if (!reader.hasNext()) {
+      return LogStorage.OP_RESULT_INVALID_ADDR;
     }
 
-    return result;
+    return address;
   }
 
   @Override
   public void close() {
     reader.close();
   }
 
+  private long readFirstBlock() {
+    // in the reader being at the end of the log
+    return findEntry(0).map(Indexed::index).orElse(LogStorage.OP_RESULT_INVALID_ADDR);
+  }
+
   /**
    * Looks up the entry whose index is either the given index, or the closest lower index.
    *
    * @param index index to seek to
    */
   public Optional<Indexed<ZeebeEntry>> findEntry(final long index) {
-    // in the future, reset/seek to the same index will be a NOOP so we can just call it all the
-    // time; right now it's a bit slow, but we will immediately implement the new journal so this
-    // is fine
-    reader.reset(index);
+    if (reader.getNextIndex() != index) {
+      final long nextIndex = reader.reset(index);
```

Comment:
RaftLogReader#reset also have checks nextIndex before reset. So we can remove it from here.

Reply:
Yes I added it after :+1: